airbyte-cdk/python/unit_tests/sources/streams/http/test_http_client.py

```diff
@@ -0,0 +1,312 @@
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+    mocked_response = MagicMock(spec=requests.Response)
+    mocked_response.status_code = 200
+    mocked_response.headers = {}
+
+    mocker.patch.object(http_client, "_create_prepared_request", return_value=prepared_request)
+    mocker.patch.object(requests.Session, "send", return_value=mocked_response)
+
+    returned_request, returned_response = http_client.send_request(http_method="get", url="https://test_base_url.com/v1/endpoint", request_kwargs={})
+
+    assert returned_request == prepared_request
+    assert returned_response == mocked_response
+
+def test_send_raises_airbyte_traced_exception_with_fail_response_action(mocker):
+    http_client = test_http_client()
+    prepared_request = requests.PreparedRequest()
+    response = requests.Response()
+    response.status_code = 400
+
+
+    mocker.patch.object(http_client._session, "send", return_value=response)
+    mocker.patch.object(http_client._error_handler, "interpret_response", return_value=ErrorResolution(ResponseAction.FAIL, FailureType.system_error, "test error message"))
+
+    with pytest.raises(AirbyteTracedException):
+        http_client._send(prepared_request, {})
+
+    assert http_client._session.send.call_count == 1
+
+def test_send_ignores_with_ignore_reponse_action_and_returns_response(mocker):
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+    mocked_response = MagicMock(spec=requests.Response)
+    mocked_response.status_code = 300
+    mocked_response.headers = {}
+    http_client._logger.info = MagicMock()
+
+    mocker.patch.object(requests.Session, "send", return_value=mocked_response)
+    mocker.patch.object(http_client._error_handler, "interpret_response", return_value=ErrorResolution(ResponseAction.IGNORE, FailureType.system_error, "test ignore message"))
+
+    returned_response = http_client._send(prepared_request, {})
+
+    http_client._logger.info.assert_called_once()
+    assert returned_response == mocked_response
+
+@pytest.mark.parametrize(
+        "backoff_time_value, exception_type",
+        [
+            (0.1, UserDefinedBackoffException),
+            (None, DefaultBackoffException)
+        ]
+)
+def test_raises_backoff_exception_with_retry_response_action(mocker, backoff_time_value, exception_type):
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+    mocked_response = MagicMock(spec=requests.Response)
+    mocked_response.status_code = 500
+    mocked_response.headers = {}
+    http_client._logger.info = MagicMock()
+
+    mocker.patch.object(http_client._backoff_strategy, "backoff_time", return_value=backoff_time_value)
```

Comment:
nit: It seems weird for me to patch internal parameters. As those are not in the public interface, those can change at any time and we wouldn't want them to generate maintenance i.e. having to update the test when the behavior does not change. In the end, we can have the same behavior by mocking the objects that we pass in `__init__`

Reply:
Fair point. Updated to mock the objects and inject them instead.